packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -1881,19 +1881,23 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
     newRoute.install(_currentOverlayEntry);
     _history[index] = newRoute;
     newRoute.didPush().whenCompleteOrCancel(() {
-      // The old route's exit is not animated. We're assuming that the
-      // new route completely obscures the old one.
       if (mounted) {
+        // The old route's exit is not animated. We're assuming that the
+        // new route completely obscures the old one.
         oldRoute
           ..didComplete(result ?? oldRoute.currentResult)
           ..dispose();
+        final int newIndex = _history.indexOf(newRoute);
```

Comment:
Is it guaranteed that newRoute is still in the history at this point? What happens if you pop it while the push animation is still ongoing? Would be good to have a test for this as well.

Reply:
In that case, newIndex is -1 and no attempt is made to call didChangeNext or didChangePrevious on transition completion. This is made slightly clearer by your suggested amendment below.
I've added a test to reproduce this use case, mainly to ensure that no exception is created. I cannot easily test the (lack of) proxy of the removed newRoute animation to the preceding route's secondaryAnimation as on completion the newRoute animation is disposed. I'm happy to amend the test if you can think of a better approach.

Reply:
I am still not quite understand what is the use case this code is trying to cover. I also tried running the test without this part of code, and it is still passing. It seems like the test does not reproduce this use case.

Reply:
When replacing a route, the 'previous' of the new route is initially the route being replaced. This should be changed once the push is complete, hence this code. The 'next' of the preceding route is also updated (though this is probably already set and unchanged a few lines down anyway). As per the discussion above, the new route may have been removed, so we check it is there first.

If you comment out the check, tests fail. If you break-point or add debug prints to the if-clause, the code is only run when appropriate. Changing the previous of  `newRoute` currently doesn't do much, and is therefore not easy to test, but it is correct to call it in case later developments make this action more important. Also note that the original code did not test  this action either - you could anyway have passed tests by removing it.

Reply:
I thought this case is covered by line 1904? where the index - 1 will be the new precedingRoute?
```dart
if (index > 0) {	  
      _history[index - 1].didChangeNext(newRoute);	
      newRoute.didChangePrevious(_history[index - 1])
}
```
or am i missing something?

Reply:
We should only notify once, either before or after. Otherwise, we might notify the precedingRoute twice if the new precedingRoute and old precedingRoute happens to be the same